File: src/Services/SurrealStore.cs
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using SurrealDb.Net;
using SurrealDb.Net.Models;

namespace OverlayCompanion.Services;

/// <summary>
/// SurrealDB-backed store for connections, settings, and audit log.
/// SurrealDB is the only database (Ryan's preferences §9). This replaces the
/// JSON-file stores in ConnectionManagementService and SettingsService with a
/// shared, durable store that both the TS management server and the C# MCP
/// server read and write. The connection string comes from the SURREALDB_*
/// environment variables (bootstrap defaults; the GUI is the source of truth).
/// </summary>
public interface ISurrealStore
{
    Task InitializeAsync(CancellationToken cancellationToken = default);
    Task<List<ConnectionConfig>> GetAllConnectionsAsync(string userId, CancellationToken cancellationToken = default);
    Task<ConnectionConfig?> GetConnectionAsync(string id, CancellationToken cancellationToken = default);
    Task<ConnectionConfig> UpsertConnectionAsync(ConnectionConfig config, CancellationToken cancellationToken = default);
    Task<bool> DeleteConnectionAsync(string id, CancellationToken cancellationToken = default);
    Task<T?> GetSettingAsync<T>(string key, T? defaultValue = default, CancellationToken cancellationToken = default);
    Task SetSettingAsync<T>(string key, T value, CancellationToken cancellationToken = default);
    Task AppendAuditAsync(AuditRecord record, CancellationToken cancellationToken = default);
}

public class SurrealStoreOptions
{
    public string Endpoint { get; set; } = "http://surrealdb:8000";
    public string Namespace { get; set; } = "overlay";
    public string Database { get; set; } = "companion";
    public string Username { get; set; } = "root";
    public string Password { get; set; } = "root";
}

public class AuditRecord
{
    public string Id { get; set; } = $"audit_log:{Guid.NewGuid():N}";
    public DateTimeOffset Timestamp { get; set; } = DateTimeOffset.UtcNow;
    public string? UserId { get; set; }
    public string Actor { get; set; } = "system";
    public string Action { get; set; } = "";
    public string? IpAddress { get; set; }
    public object? Detail { get; set; }
}

/// <summary>
/// Record model for a saved connection row in SurrealDB. Implements IRecord so
/// the SDK's Upsert/Create generics accept it. The Id is a RecordId built from
/// the existing Guid-based id so prior data keeps working.
/// </summary>
public class SurrealConnectionRecord : IRecord
{
    public RecordId Id { get; set; }
    public string Name { get; set; } = string.Empty;
    public string Host { get; set; } = string.Empty;
    public int Port { get; set; }
    public string Protocol { get; set; } = "kasmvnc";
    public string? Username { get; set; }
    public bool Active { get; set; }
    public DateTimeOffset CreatedAt { get; set; } = DateTimeOffset.UtcNow;
    public DateTimeOffset UpdatedAt { get; set; } = DateTimeOffset.UtcNow;
    public DateTimeOffset? LastConnected { get; set; }
}

public class SurrealSettingRecord : IRecord
{
    public RecordId Id { get; set; }
    public object Value { get; set; } = new { };
    public string Category { get; set; } = "general";
    public DateTimeOffset UpdatedAt { get; set; } = DateTimeOffset.UtcNow;
}

public class SurrealAuditRecord : IRecord
{
    public RecordId Id { get; set; }
    public DateTimeOffset Timestamp { get; set; } = DateTimeOffset.UtcNow;
    public string? UserId { get; set; }
    public string Actor { get; set; } = "system";
    public string Action { get; set; } = "";
    public string? IpAddress { get; set; }
    public object? Detail { get; set; }
}

public class SurrealStore : ISurrealStore
{
    private readonly ILogger<SurrealStore> _logger;
    private readonly SurrealStoreOptions _options;
    private readonly ISurrealDbClient _db;
    private int _initState; // 0=uninit, 1=initializing, 2=done

    public SurrealStore(ILogger<SurrealStore> logger, IOptions<SurrealStoreOptions> options)
    {
        _logger = logger;
        _options = options.Value;
        // SurrealDbClient is singleton-safe per the SDK lifetime table. The
        // connection auto-reconnects when the DB restarts.
        _db = new SurrealDbClient(_options.Endpoint);
    }

    public async Task InitializeAsync(CancellationToken cancellationToken = default)
    {
        if (Interlocked.CompareExchange(ref _initState, 1, 0) != 0) return;
        try
        {
            await _db.Use(_options.Namespace, _options.Database, cancellationToken);
            await _db.SignIn(new SurrealDb.Net.Models.Auth.RootAuth
            {
                Username = _options.Username,
                Password = _options.Password,
            }, cancellationToken);
            _initState = 2;
            _logger.LogInformation("SurrealDB store connected to {Endpoint} (ns={Ns}, db={Db})",
                _options.Endpoint, _options.Namespace, _options.Database);
        }
        catch (Exception ex)
        {
            _initState = 0;
            // Non-fatal: the file-based fallbacks in the services keep the app
            // working without a DB. The health endpoint reports reachability.
            _logger.LogWarning(ex, "SurrealDB connection failed; services will use file fallback");
        }
    }

    private bool IsReady => _initState == 2;

    public async Task<List<ConnectionConfig>> GetAllConnectionsAsync(string userId, CancellationToken cancellationToken = default)
    {
        if (!IsReady) return new List<ConnectionConfig>();
        try
        {
            var rows = await _db.Select<SurrealConnectionRecord>("connection", cancellationToken);
            return rows.Select(ToConfig).ToList();
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "SurrealDB connection select failed");
            return new List<ConnectionConfig>();
        }
    }

    public async Task<ConnectionConfig?> GetConnectionAsync(string id, CancellationToken cancellationToken = default)
    {
        if (!IsReady) return null;
        try
        {
            var recordId = ParseRecordId(id, "connection");
            var row = await _db.Select<SurrealConnectionRecord>(recordId, cancellationToken);
            return row == null ? null : ToConfig(row);
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "SurrealDB connection get failed for {Id}", id);
            return null;
        }
    }

    public async Task<ConnectionConfig> UpsertConnectionAsync(ConnectionConfig config, CancellationToken cancellationToken = default)
    {
        if (!IsReady)
        {
            return config;
        }
        try
        {
            var record = FromConfig(config);
            var saved = await _db.Upsert(record, cancellationToken);
            return ToConfig(saved);
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "SurrealDB connection upsert failed");
            return config;
        }
    }

    public async Task<bool> DeleteConnectionAsync(string id, CancellationToken cancellationToken = default)
    {
        if (!IsReady) return false;
        try
        {
            var recordId = ParseRecordId(id, "connection");
            return await _db.Delete(recordId, cancellationToken);
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "SurrealDB connection delete failed for {Id}", id);
            return false;
        }
    }

    public async Task<T?> GetSettingAsync<T>(string key, T? defaultValue = default, CancellationToken cancellationToken = default)
    {
        if (!IsReady) return defaultValue;
        try
        {
            var recordId = ParseRecordId(key, "app_config");
            var row = await _db.Select<SurrealSettingRecord>(recordId, cancellationToken);
            if (row == null) return defaultValue;
            return System.Text.Json.JsonSerializer.Deserialize<T>(row.Value.ToString() ?? "");
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "SurrealDB setting get failed for {Key}", key);
            return defaultValue;
        }
    }

    public async Task SetSettingAsync<T>(string key, T value, CancellationToken cancellationToken = default)
    {
        if (!IsReady) return;
        try
        {
            var recordId = ParseRecordId(key, "app_config");
            var record = new SurrealSettingRecord
            {
                Id = recordId,
                Value = value!,
                Category = key.Contains('.') ? key.Split('.')[0] : "general",
                UpdatedAt = DateTimeOffset.UtcNow,
            };
            await _db.Upsert(record, cancellationToken);
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "SurrealDB setting set failed for {Key}", key);
        }
    }

    public async Task AppendAuditAsync(AuditRecord record, CancellationToken cancellationToken = default)
    {
        if (!IsReady) return;
        try
        {
            var row = new SurrealAuditRecord
            {
                Id = ParseRecordId(record.Id, "audit_log"),
                Timestamp = record.Timestamp,
                UserId = record.UserId,
                Actor = record.Actor,
                Action = record.Action,
                IpAddress = record.IpAddress,
                Detail = record.Detail,
            };
            await _db.Create(row, cancellationToken);
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "SurrealDB audit append failed");
        }
    }

    private static RecordId ParseRecordId(string id, string table)
    {
        // Accept either "table:id" or a bare id; normalize to a RecordId.
        var parts = id.Split(':', 2);
        if (parts.Length == 2)
        {
            return RecordId.From(parts[0], parts[1]);
        }
        return RecordId.From(table, id);
    }

    private static SurrealConnectionRecord FromConfig(ConnectionConfig config)
    {
        var id = ParseRecordId(config.Id, "connection");
        return new SurrealConnectionRecord
        {
            Id = id,
            Name = config.Name,
            Host = config.Host,
            Port = config.Port,
            Protocol = config.Protocol,
            Username = config.Username,
            Active = config.IsActive,
            CreatedAt = config.CreatedAt,
            UpdatedAt = DateTime.UtcNow,
            LastConnected = config.LastConnected,
        };
    }

    private static ConnectionConfig ToConfig(SurrealConnectionRecord row)
    {
        return new ConnectionConfig
        {
            Id = row.Id.ToString(),
            Name = row.Name,
            Host = row.Host,
            Port = row.Port,
            Protocol = row.Protocol,
            Username = row.Username,
            IsActive = row.Active,
            CreatedAt = row.CreatedAt.UtcDateTime,
            LastConnected = row.LastConnected?.UtcDateTime,
        };
    }
}

File: infra/server/src/surreal-store.ts
/**
 * SurrealDB store for the management server.
 *
 * SurrealDB is the only database (Ryan's preferences §9). This module talks
 * to SurrealDB over its HTTP /sql endpoint — no JS SDK dependency, keeping the
 * dependency surface minimal. The connection details come from the
 * SURREALDB_* environment variables (bootstrap defaults; the GUI is the source
 * of truth for runtime config).
 *
 * All queries are parameterized to avoid injection. The response envelope from
 * /sql is an array of { result, status, time }; this helper unwraps the first
 * statement's result by default and throws on a non-OK status.
 */

export interface SurrealDbOptions {
  endpoint: string;
  namespace: string;
  database: string;
  username: string;
  password: string;
}

export interface AuditEvent {
  action: string;
  userId?: string;
  actor?: string;
  ipAddress?: string;
  detail?: Record<string, unknown>;
  traceId?: string;
}

function envOrDefault(name: string, fallback: string): string {
  const v = process.env[name];
  return v && v.length > 0 ? v : fallback;
}

export function loadSurrealOptions(): SurrealDbOptions {
  return {
    endpoint: envOrDefault('SURREALDB_URL', 'http://surrealdb:8000'),
    namespace: envOrDefault('SURREALDB_NAMESPACE', 'overlay'),
    database: envOrDefault('SURREALDB_DATABASE', 'companion'),
    username: envOrDefault('SURREALDB_USERNAME', 'root'),
    password: envOrDefault('SURREALDB_PASSWORD', 'root'),
  };
}

interface SqlResponseRow {
  result?: unknown;
  status?: string;
  detail?: string;
}

export class SurrealDbStore {
  private readonly opts: SurrealDbOptions;
  private readonly authHeader: string;
  // Lazy schema-apply guard so we only run the DEFINE statements once per process.
  private schemaApplied = false;
  private schemaApplyPromise: Promise<void> | null = null;

  constructor(opts: SurrealDbOptions = loadSurrealOptions()) {
    this.opts = opts;
    this.authHeader =
      'Basic ' + Buffer.from(`${opts.username}:${opts.password}`).toString('base64');
  }

  /**
   * Run one or more SurrealQL statements. Variables are bound positionally.
   * Returns the result of the first statement by default; pass index to pick
   * another. Throws on any statement returning a non-OK status.
   */
  async query<T = unknown>(sql: string, vars?: Record<string, unknown>, resultIndex = 0): Promise<T> {
    const body = vars ? JSON.stringify(vars) : '';
    const res = await fetch(`${this.opts.endpoint}/sql`, {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
        'Surreal-NS': this.opts.namespace,
        'Surreal-DB': this.opts.database,
        Authorization: this.authHeader,
      },
      body: vars ? this.interpolateVars(sql, vars) : sql,
      signal: AbortSignal.timeout(10000),
    });

    if (!res.ok) {
      const text = await res.text().catch(() => '');
      throw new Error(`SurrealDB HTTP ${res.status}: ${text.slice(0, 200)}`);
    }

    const rows = (await res.json()) as SqlResponseRow[];
    if (!Array.isArray(rows)) {
      throw new Error('SurrealDB returned non-array response');
    }

    const row = rows[resultIndex];
    if (!row) {
      throw new Error(`SurrealDB response missing index ${resultIndex}`);
    }
    if (row.status && row.status !== 'OK') {
      throw new Error(`SurrealDB query error: ${row.detail ?? row.status}`);
    }
    return row.result as T;
  }

  /**
   * SurrealDB HTTP /sql does not bind JSON body variables directly in v2; we
   * interpolate by safely serializing each var. Strings are escaped; objects
   * are rendered as SurrealQL objects. This is bounded to our own queries.
   */
  private interpolateVars(sql: string, vars: Record<string, unknown>): string {
    let out = sql;
    for (const [key, value] of Object.entries(vars)) {
      const token = `$${key}`;
      out = out.split(token).join(this.renderValue(value));
    }
    return out;
  }

  private renderValue(value: unknown): string {
    if (value === null || value === undefined) return 'NONE';
    if (typeof value === 'string') return JSON.stringify(value);
    if (typeof value === 'number' || typeof value === 'boolean') return String(value);
    return JSON.stringify(value);
  }

  /**
   * Apply the schema file on boot. Idempotent (OVERWRITE). Safe to call
   * repeatedly; only runs once per process.
   */
  async ensureSchema(schemaSql: string): Promise<void> {
    if (this.schemaApplied) return;
    if (this.schemaApplyPromise) return this.schemaApplyPromise;
    this.schemaApplyPromise = (async () => {
      try {
        await fetch(`${this.opts.endpoint}/sql`, {
          method: 'POST',
          headers: {
            Accept: 'application/json',
            'Content-Type': 'text/plain',
            'Surreal-NS': this.opts.namespace,
            'Surreal-DB': this.opts.database,
            Authorization: this.authHeader,
          },
          body: schemaSql,
          signal: AbortSignal.timeout(15000),
        });
        this.schemaApplied = true;
      } catch (err) {
        // Schema apply failure is non-fatal at boot; queries will surface a
        // clearer error. The health check reports DB reachability.
        console.warn('[WARN] SurrealDB schema apply failed:', (err as Error).message);
      }
    })();
    return this.schemaApplyPromise;
  }

  /**
   * Lightweight reachability probe for the health endpoint.
   */
  async ping(): Promise<boolean> {
    try {
      const res = await fetch(`${this.opts.endpoint}/health`, {
        signal: AbortSignal.timeout(3000),
      });
      return res.ok;
    } catch {
      return false;
    }
  }

  // ---- User records -----------------------------------------------------

  async upsertUser(user: {
    id?: string;
    provider: string;
    subject?: string;
    username: string;
    email?: string;
    displayName?: string;
    roles?: string[];
    passwordHash?: string;
  }): Promise<{ id: string }> {
    const id = user.id ?? `user:${cryptoRandomId()}`;
    // Local-auth users are looked up by username (never by IdP subject), but
    // the unique (provider, subject) index would collide for every local user
    // (subject=NONE). Give locals a unique per-user subject sentinel so the
    // subject index stays meaningful for OIDC without breaking multi-local
    // registration. OIDC callers still provide their real subject.
    const subject = user.subject ?? (user.provider === 'local' ? `local:${id}` : null);
    await this.query(
      `UPSERT type::thing($id) SET
        provider = $provider,
        subject = $subject,
        username = $username,
        email = $email,
        display_name = $displayName,
        roles = $roles,
        password_hash = $passwordHash,
        updated_at = time::now(),
        last_login = time::now();`,
      {
        id,
        provider: user.provider,
        subject,
        username: user.username,
        email: user.email ?? null,
        displayName: user.displayName ?? null,
        roles: user.roles ?? [],
        passwordHash: user.passwordHash ?? null,
      },
    );
    return { id };
  }

  async findUserBySubject(provider: string, subject: string): Promise<DbUser | null> {
    const rows = await this.query<DbUser[]>(
      `SELECT * FROM user WHERE provider = $provider AND subject = $subject LIMIT 1;`,
      { provider, subject },
    );
    return rows && rows.length > 0 ? rows[0] : null;
  }

  async findUserByUsername(username: string): Promise<DbUser | null> {
    const rows = await this.query<DbUser[]>(
      `SELECT * FROM user WHERE username = $username LIMIT 1;`,
      { username },
    );
    return rows && rows.length > 0 ? rows[0] : null;
  }

  async getUser(id: string): Promise<DbUser | null> {
    const rows = await this.query<DbUser[]>(`SELECT * FROM type::thing($id);`, { id });
    return rows && rows.length > 0 ? rows[0] : null;
  }

  async deleteUser(id: string): Promise<void> {
    await this.query(`DELETE FROM type::thing($id);`, { id });
  }

  async countUsers(): Promise<number> {
    const rows = await this.query<Array<{ count: number }>>(
      `SELECT count() AS count FROM user GROUP ALL;`,
    );
    if (Array.isArray(rows) && rows.length > 0) {
      return rows[0].count ?? 0;
    }
    const single = rows as unknown as { count?: number };
    return single?.count ?? 0;
  }

  // ---- Sessions ---------------------------------------------------------

  async createSession(session: {
    userId: string;
    tokenHash: string;
    csrfToken?: string;
    ipAddress?: string;
    userAgent?: string;
    expiresAt: Date;
  }): Promise<string> {
    const id = `session:${cryptoRandomId()}`;
    await this.query(
      `UPSERT type::thing($id) SET
        user_id = type::thing($userId),
        token_hash = $tokenHash,
        csrf_token = $csrfToken,
        ip_address = $ipAddress,
        user_agent = $userAgent,
        expires_at = type::datetime($expiresAt),
        created_at = time::now(),
        revoked = false;`,
      {
        id,
        userId: session.userId,
        tokenHash: session.tokenHash,
        csrfToken: session.csrfToken ?? null,
        ipAddress: session.ipAddress ?? null,
        userAgent: session.userAgent ?? null,
        expiresAt: session.expiresAt.toISOString(),
      },
    );
    return id;
  }

  async findSessionByTokenHash(tokenHash: string): Promise<DbSession | null> {
    const rows = await this.query<DbSession[]>(
      `SELECT * FROM session WHERE token_hash = $tokenHash AND revoked = false LIMIT 1;`,
      { tokenHash },
    );
    return rows && rows.length > 0 ? rows[0] : null;
  }

  async revokeSession(id: string): Promise<void> {
    await this.query(`UPDATE type::thing($id) SET revoked = true;`, { id });
  }

  async revokeAllUserSessions(userId: string): Promise<void> {
    await this.query(
      `UPDATE session SET revoked = true WHERE user_id = type::thing($userId);`,
      { userId },
    );
  }

  // ---- Audit log --------------------------------------------------------

  async appendAudit(event: AuditEvent): Promise<void> {
    const id = `audit_log:${cryptoRandomId()}`;
    // user_id is an option<record<user>>. When a userId is present it must be
    // cast with type::thing to satisfy the record type; when absent it stays
    // NONE (type::thing(NONE) would error).
    const userIdClause = event.userId
      ? `type::thing($userId)`
      : `NONE`;
    const vars: Record<string, unknown> = {
      id,
      actor: event.actor ?? 'system',
      action: event.action,
      ipAddress: event.ipAddress ?? null,
      detail: event.detail ?? null,
      traceId: event.traceId ?? null,
    };
    if (event.userId) vars.userId = event.userId;
    await this.query(
      `UPSERT type::thing($id) SET
        timestamp = time::now(),
        user_id = ${userIdClause},
        actor = $actor,
        action = $action,
        ip_address = $ipAddress,
        detail = $detail,
        trace_id = $traceId;`,
      vars,
    );
  }

  // ---- App configuration (GUI-first) ------------------------------------

  // The dotted key (e.g. 'auth.oidc') cannot be a SurrealDB record id directly
  // ('.' is the field separator), so the record id is a dot-free encoded form
  // and the canonical dotted key is stored in `name`. Lookups use `name`.
  private static configRecordId(key: string): string {
    return 'app_config:' + Buffer.from(key).toString('base64url');
  }

  async getConfig(key: string): Promise<unknown | null> {
    const rows = await this.query<Array<{ payload: string }>>(
      'SELECT `payload` FROM app_config WHERE `name` = $name LIMIT 1;',
      { name: key },
    );
    if (!rows || rows.length === 0) return null;
    return this.parsePayload(rows[0].payload);
  }

  // payload is stored as a JSON string (SurrealQL can't take arbitrary JSON
  // object literals via string interpolation). Parse back to an object.
  private parsePayload(payload: string): unknown {
    if (payload == null) return null;
    try {
      return JSON.parse(payload);
    } catch {
      return payload;
    }
  }

  async setConfig(key: string, value: Record<string, unknown>, category: string, updatedBy?: string): Promise<void> {
    const vars: Record<string, unknown> = {
      id: SurrealDbStore.configRecordId(key),
      name: key,
      value: JSON.stringify(value),
      category,
    };
    const updatedByClause = updatedBy ? `type::thing($updatedBy)` : `NONE`;
    if (updatedBy) vars.updatedBy = updatedBy;
    await this.query(
      `UPSERT type::thing($id) SET
        \`name\` = $name,
        \`payload\` = $value,
        category = $category,
        updated_at = time::now(),
        updated_by = ${updatedByClause};`,
      vars,
    );
  }

  async getConfigByCategory(category: string): Promise<Array<{ id: string; value: unknown }>> {
    const rows = await this.query<Array<{ id: string; name: string; payload: string }>>(
      'SELECT id, `name`, `payload` FROM app_config WHERE category = $category;',
      { category },
    );
    // The app reads the canonical dotted key from `name`; `id` is the encoded
    // record id. `value` here is the settings payload (JSON string in the DB,
    // parsed back to an object for the app).
    return (rows ?? []).map((r) => ({ id: r.name ?? r.id, value: this.parsePayload(r.payload) }));
  }
  // ---- Connections -------------------------------------------------------

  async listConnections(userId: string): Promise<DbConnection[]> {
    const rows = await this.query<DbConnection[]>(
      'SELECT * FROM connection WHERE user_id = type::thing($userId) ORDER BY created_at DESC;',
      { userId: `user:${userId.replace(/^user:/, '')}` },
    );
    return rows ?? [];
  }

  async getConnection(userId: string, id: string): Promise<DbConnection | null> {
    const rows = await this.query<DbConnection[]>(
      'SELECT * FROM connection WHERE id = type::thing($id) AND user_id = type::thing($userId) LIMIT 1;',
      { id: `connection:${id.replace(/^connection:/, '')}`, userId: `user:${userId.replace(/^user:/, '')}` },
    );
    return rows && rows.length > 0 ? rows[0] : null;
  }

  async upsertConnection(userId: string, input: ConnectionInput, passwordHash?: string): Promise<DbConnection> {
    const id = input.id ?? `connection:${cryptoRandomId()}`;
    const cleanId = id.replace(/^connection:/, '');
    const vars: Record<string, unknown> = {
      id: `connection:${cleanId}`,
      userId: `user:${userId.replace(/^user:/, '')}`,
      name: input.name,
      host: input.host,
      port: input.port,
      protocol: input.protocol,
      username: input.username ?? null,
      ssl: input.ssl ?? false,
      description: input.description ?? null,
      passwordHash: passwordHash ?? null,
    };
    await this.query(
      `UPSERT type::thing($id) SET
        user_id = type::thing($userId),
        \`name\` = $name,
        host = $host,
        port = $port,
        protocol = $protocol,
        username = $username,
        password_hash = $passwordHash,
        ssl = $ssl,
        \`description\` = $description,
        updated_at = time::now();`,
      vars,
    );
    const saved = await this.getConnection(userId, cleanId);
    if (!saved) throw new Error('Connection upsert failed');
    return saved;
  }

  async deleteConnection(userId: string, id: string): Promise<boolean> {
    const cleanId = id.replace(/^connection:/, '');
    const rows = await this.query<DbConnection[]>(
      'DELETE type::thing($id) WHERE user_id = type::thing($userId) RETURN BEFORE;',
      { id: `connection:${cleanId}`, userId: `user:${userId.replace(/^user:/, '')}` },
    );
    return rows !== null && rows.length > 0;
  }

  async touchLastConnected(userId: string, id: string): Promise<void> {
    const cleanId = id.replace(/^connection:/, '');
    await this.query(
      'UPDATE type::thing($id) SET last_connected = time::now(), updated_at = time::now() WHERE user_id = type::thing($userId);',
      { id: `connection:${cleanId}`, userId: `user:${userId.replace(/^user:/, '')}` },
    );
  }
}

export interface DbUser {
  id: string;
  provider: string;
  subject?: string;
  username: string;
  email?: string;
  display_name?: string;
  roles: string[];
  active: boolean;
  created_at?: string;
  updated_at?: string;
  last_login?: string;
  password_hash?: string;
}

export interface DbSession {
  id: string;
  user_id: string;
  token_hash: string;
  csrf_token?: string;
  ip_address?: string;
  user_agent?: string;
  created_at?: string;
  expires_at: string;
  revoked: boolean;
}

// ---- Connections ---------------------------------------------------------

/** A saved remote-desktop connection row. Mirrors the `connection` table. */
export interface DbConnection {
  id: string;
  user_id: string;
  name: string;
  host: string;
  port: number;
  protocol: string; // 'kasmvnc' | 'vnc' | 'rdp'
  username?: string | null;
  password_hash?: string | null;
  ssl?: boolean;
  description?: string | null;
  active?: boolean;
  created_at?: string;
  updated_at?: string;
  last_connected?: string | null;
}

/** Data needed to create or update a connection (no password hash here). */
export interface ConnectionInput {
  id?: string;
  name: string;
  host: string;
  port: number;
  protocol: string;
  username?: string | null;
  ssl?: boolean;
  description?: string | null;
}

function cryptoRandomId(): string {
  return crypto.randomUUID().replace(/-/g, '');
}
